src/node_api.h

```diff
@@ -14,21 +14,18 @@
 #include <stdbool.h>
 #include "node_api_types.h"
 
-#ifdef _WIN32
-  #ifdef BUILDING_NODE_EXTENSION
-    #ifdef EXTERNAL_NAPI
-      // Building external N-API, or native module against external N-API
-      #define NAPI_EXTERN /* nothing */
+#ifndef NAPI_EXTERN
```

Comment:
I don't see the point of letting users override it.  The definition needs to reflect how node_api.cc is compiled, otherwise linking won't work.  You can't override `NODE_EXTERN`, `UV_EXTERN` or `V8_EXPORT` either.

Reply:
Well, it helps us out in node-addon-api by letting us do this:
```gyp
      'conditions': [
        [ "OS=='win'", {
          'defines': ['NAPI_EXTERN=']
        }, {
          'defines': ['NAPI_EXTERN=__attribute__((visibility("hidden")))']
        } ]
      ]
```
to prevent conflicts with Node.js between versions >= 8.0.0 and < 8.6.0.

Reply:
... where both node-addon-api and Node.js provide the same symbols, but node-addon-api's symbols must be preferred.

Reply:
But that's the wrong way to go about it.  The right way is to fix the node-addon-api build so that the symbols from the static library are searched before symbols from the node binary.

This is in the context of https://github.com/nodejs/node-addon-api/issues/142, right?  If you look at the [nm -C gist](https://gist.github.com/gabrielschulhof/0fe4442360f119facc48654b6c3710f9) you posted, there are a lot of n-api symbols in there that are weak (W) when they shouldn't be.

Reply:
Actually, those symbols belong to the C++ wrapper, and none of them overlap with Node.js. None of the symbols which overlap with Node.js are weak.

I could find no way to cause the symbols from the static library to be searched before the symbols from Node.js, because without this change they are both declared with default visibility. In fact, visibility hidden is explicitly designed to indicate that, although the symbols are declared global because they have to be in different compilation units, they are not intended for consumption outside the .so and are thus to be treated, in a way, as static.

Reply:
OK, NM. I managed to get `-fvisibility=hidden` to work for node-addon-api. I guess you re-motivated me to try again :) Thanks!

Reply:
> those symbols belong to the C++ wrapper, and none of them overlap with Node.js

Still, they probably shouldn't be weak.  Different versions might clash and, if nothing else, it means the dynamic linker has to do more work when the module is loaded.  Might be worth adding a regression test for in node-addon-api.